config/reader.go

```diff
@@ -327,14 +463,11 @@ func fromReader(
 	shouldReadFromCloud bool,
 ) (*Config, error) {
 	// First read and processes config from disk
-	unprocessedConfig := Config{
-		ConfigFilePath: originalPath,
-	}
-	err := json.NewDecoder(r).Decode(&unprocessedConfig)
+	unprocessedConfig, err := GenerateConfigFromFile(originalPath)
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to decode Config from json")
+		return nil, err
 	}
-	cfgFromDisk, err := processConfigLocalConfig(&unprocessedConfig, logger)
+	cfgFromDisk, err := processConfigLocalConfig(unprocessedConfig, logger)
```

Comment:
[design question] 

Is there a reason to prefer `GenerateConfigFromFile` over adding something to the start of `reader.go::processConfig(...)` that does the replacement? That function seems to do other modifications similar to what we are doing.

Reply:
yeah so there is definitely room for debate where we add this in. The reason that I do before we pass in the config was to minimize the changes to the reader code so far. We want to change the placeholders before we parse this as a config so to do that we would then need to change all the definitions for these functions to be files/readers rather than the configs themselves. We already convert the file to the config before we pass it in here anyways, so it made sense to me to wrap only that function than the actual process config function.

Reply:
Yeah we'll want to replace at a higher level than at this level. Otherwise we'll have to traverse the config (the walker approach that existed before). So vote to keep the replacement logic where it is in the this PR.